packages/flutter/lib/src/cupertino/tab.dart

```diff
@@ -0,0 +1,161 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+import 'route.dart';
+
+/// A single tab with its own [Navigator] state and history.
+///
+/// A typical tab used as the content of each tab in a [CupertinoTabScaffold]
+/// where multiple tabs with parallel navigation states and history can
+/// co-exist.
+///
+/// [CupertinoTab] configures the top-level [Navigator] to search for routes
+/// in the following order:
+///
+///  1. For the `/` route, the [home] property, if non-null, is used.
+///
+///  2. Otherwise, the [routes] table is used, if it has an entry for the route,
+///     including `/` if [home] is not specified.
+///
+///  3. Otherwise, [onGenerateRoute] is called, if provided. It should return a
+///     non-null value for any _valid_ route not handled by [home] and [routes].
+///
+///  4. Finally if all else fails [onUnknownRoute] is called.
+///
+/// These navigation properties is not shared with any sibling [CupertinoTab]
+/// nor any ancestor or descendent [Navigator] instances.
+///
+/// Be sure to acquire a [BuildContext] that belongs to the [CupertinoTab] (such
+/// as by using a [Builder]) to ensure that [Navigator.of] calls uses this
+/// [CupertinoTab]'s [Navigator] rather than that of its parent.
```

Comment:
this sentence doesn't say when this is important. Be sure to do this... when?

Reply:
Actually on more thoughts, this thing might be a trap waiting to spring. Maybe I should just let the API be a Builder instead of a Widget. Any objections?

Reply:
You should consider it a rule that any time you allow a widget to have a child, it should accept any Widget at all. The way that one of the widgets here requires a CupertinoTabBar is already sketchy.

Flutter is built on the idea of unopinionated composition, meaning you can build things out of other things, and they just all fit together. The least satisfying lego bricks are the ones that only fit one other kind of lego piece. :-)

Reply:
I think we crossed threads somewhere. 

For the whole thing about making sure to use a builder to reference the right navigator, I just avoided that trap and made the user give a builder instead of a widget.  (the builder will take any widget returned if that's what you meant)

For being able to specify any widgets, I agree. Will leave #12078 open to let the tabBar property be anything.